### DISCUSSION TITLE
Hvilke nettlesere og hjelpemidler tester vi med?

### DISCUSSION THREAD

**Discussion by mrosvik** [digdir/designsystemet#2908] [Feature requests]
## Utfordring
Hvilke nettlesere og hjelpemidler tester vi med? Kunne det vært en felles anbefaling på tvers i det offentlige?

## Mål med tråden
Samle info om hva ulike etater tester på og hvilke krav de legger til grunn. Ut i fra innsikten kan vi kanskje vurdere om det gir mening med noe felles anbefaling.

## Relevant innsikt
- Flere definerer støtten sin basert på [caniuse.com](https://caniuse.com/) "we currently support 98% of all users on Chrome, Edge, Firefox, Firefox for Android, Safari, Safari on iOS, Samsung Internet, and Opera." [Eksempel fra u-elements](https://u-elements.github.io/u-elements/guide/support).
- Brønnøysundregistrene har noen testprosedyrer som sikrer at en tester på nettlesere med ulike motorer, så en trenger ikke teste på alle nettleserne, men på alle motorene.  [Info under](https://github.com/digdir/designsystemet/discussions/2908#discussioncomment-11581657). 
- [Skatteetaten](https://www.skatteetaten.no/stilogtone/designsystemet/kom-i-gang/nettleserstotte/) definerer det slik: "Designsystemet støtter nettlesere som blir brukt av mer enn 2 % av besøkene på skatteetaten.no"
- Nav: "[Hvilke nettlesere, enheter og hjelpemidler støtter vi?](https://aksel.nav.no/god-praksis/artikler/nettleserstotte)"
- GovUK - [Browser and assistive technology support](https://github.com/alphagov/govuk-frontend#browser-and-assistive-technology-support)
- UU-tilsynet anbefaler at det testes i: 
   - De vanligste moderne nettleserne (som Chrome, Firefox, Safari, Microsoft Edge). 
   - Forskjellige operativsystemer (Windows, macOS, iOS, Android). 
   - Forskjellige skjermlesere og andre hjelpemidler

**Comment by Camulos** [comment]
Vi (Brønnøysundregistrene) har 2 testprosedyrer som handler om nettlesere, krav 801 og 802 i kravsettet for brukskvalitet. I tillegg har vi et krav om at det skal være testet med VO og Jaws.
Gjengitt nedenfor slik de er beskrevet i rutinedokumentet:

**801. Løsningen fungerer tilfredsstillende i de mest
brukte nettlesere.**

1. Test med disse på Windows: 
- Chrome, Edge eller Opera (Blink/Chromium)
- Firefox (Gecko)
2. Test med disse på MacOS:
- Safari (Webkit) – husk å skru på støtte for tabnavigasjon
- Chrome, Edge eller Opera (Blink/Chromium)
- Firefox (Gecko) – husk å skru på støtte for tabnavigasjon

Test på minst en nettleser med hver type nettlesermotor,
 fungerer noe ikke tilfredsstillende med en av dem skal kravet feile. Forsøk å
 utføre hovedoppgaven til løsningen i tillegg til de vanligste sekundære
 oppgavene. Forsøk også med testaturnavigering (tabnavigering). Noter hvilke
 nettlesere som feilet og beskriv opplevde feil. 

Krysssjekk med funn på krav 502 (WCAG 2.1: 2.1.1 Tastatur) og 504 (WCAG 2.1: 2.1.2 Ingen tastaturfeller) i tilfelle feil på
 disse viser seg å være nettleserspesifikke, kan hende det da må opprettes
 feilrapport hos den respektive utvikleren av nettleseren. Finnes det en kjent
 feil kan kravet gis «godkjent med kommentar» med henvisning til dokumentert
 feil hos nettleserprodusenten.


**802. Løsningen fungerer tilfredsstillende i de mest brukte nettleserne på nettbrett og mobil med iOS og Android (reelle, ikke emulerte, enheter).**

1. Test med disse på Android:
- Samsung Internet (Blink/Chromium)
- Opera Mobile eller Chrome (Blink/Chromium)
- Firefox (Gecko)
2. Test med disse på iOS:
- Safari (Webkit)
- Opera Mobile, Firefox eller Chrome (Webkit)

Test på minst en nettleser med hver type nettlesermotor og minst en konkurrent av plattformens standard nettleser også når alle nettlesere på plattformen bygges på samme nettlesermotor. Fungerer noe ikke tilfredsstillende med en av dem skal kravet feile. Noter hvilke nettlesere som feilet og beskriv opplevde feil.

**Comment by frankdahle** [comment]
### Her er referat fra en diskusjon på Slack om nettleserstøtte:

**Problemstillinger**
**- Manglende dokumentasjon:** Det finnes ingen offisiell oversikt over hvilke nettleserversjoner designsystemet støtter.
**- Kompatibilitet med eldre enheter:** Eldre iOS-enheter (som ikke kan oppdateres til nyere Safari-versjoner) skaper problemer, spesielt med funksjoner som popover og CSS-funksjoner som @layer og :has.
**- Statistikk vs. realitet:** Selv om statistikk viser lav bruk av gamle nettlesere, oppstår mange feilrapporter fra brukere med eldre enheter.
**- Safari på iOS:** Nettleseren er bundet til OS-versjonen, og alternative nettlesere er i praksis Safari med skall.
**- Global bruk:** Fokus på >2 % i Norge kan utelukke brukere fra utlandet.
**- Polyfill-håndtering:** Uenighet om polyfills skal inkluderes i komponentene eller overlates til brukerne.
**- Bevegelig mål:** Vanskelig å definere en presis og vedlikeholdsfri standard for støtte.

**Løsningsforslag**
**Browserslist-konfigurasjon:** Nåværende praksis er å støtte 90 % dekning i Norge, siste 2 versjoner, Firefox ESR, og ekskludere utdaterte nettlesere (IE ≤ 11, Opera Mini).

**Polyfills:**
- Bruk polyfills for moderne funksjoner som popover der nødvendig. Men bør legges på av konsument
- Alternativ: veilede brukere om hvordan polyfills kan implementeres.
- Nasjonale retningslinjer: Forslag om samarbeid for å etablere felles retningslinjer for nettleserstøtte.

**Pragmatisk tilnærming:**
- Følge praksis fra Skatteetaten og NAV – støtte alle nettlesere som brukes av mer enn 2 % av sluttbrukerne, med forbehold.
- Vurdere global bruk der relevant.

**Kommunikasjon:**
- Dokumentere målsetting som «Vi tar sikte på å støtte …» fremfor garantier.
- Beskrive forbehold i dokumentasjon og blogginnlegg.
- Vurdere varsel til brukere med utdaterte nettlesere for å redusere feil og klager.

**Videre tiltak:**
- Diskutere polyfill-strategi i teamet.
- Vurdere foredrag eller opplæring om polyfills og Browserslist.
- Utforske nye Browserslist-funksjoner som Baseline for enklere vedlikehold.

**Lenker referert til i diskusjonen:**
https://caniuse.com/?search=popover
https://github.com/oddbird/popover-polyfill
https://gs.statcounter.com/browser-market-share/all/norway
https://www.skatteetaten.no/stilogtone/designsystemet/kom-i-gang/nettleserstotte/
https://aksel.nav.no/god-praksis/artikler/nettleserstotte
https://web.dev/blog/browserslist-supports-baseline
